### PR TITLE
Add resctrl metrics.

### PR DESCRIPTION
**Pull request by Creatone**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -30,31 +26,25 @@ require (
...
 	github.com/mistifyio/go-zfs v2.1.2-0.20190413222219-f784269be439+incompatible
 	github.com/morikuni/aec v1.0.0 // indirect
-	github.com/mrunalp/fileutils v0.0.0-20160930181131-4ee1cc9a8058 // indirect
 	github.com/opencontainers/go-digest v1.0.0-rc1 // indirect
 	github.com/opencontainers/image-spec v1.0.1 // indirect
 	github.com/opencontainers/runc v1.0.0-rc10
 	github.com/opencontainers/runtime-spec v1.0.2
-	github.com/opencontainers/selinux v1.3.3 // indirect
-	github.com/pkg/errors v0.8.1
+	github.com/pkg/errors v0.9.1
```
Comment (iwankgb) [inline go.mod:35]: Do we need to introduce new dependency?

File: cmd/internal/container/mesos/handler.go [hunk 2]
```diff
@@ -72,6 +74,20 @@ func newMesosContainerHandler(
 		Paths: cgroupPaths,
 	}
 
+	var resctrlManager *resctrl.IntelRdtManager
```
Comment (iwankgb) [inline cmd/internal/container/mesos/handler.go]: This is `nil` pointer and possible source of SEGFAULT. Consider doing what I did in `stats` package and create no-op implementation of `intelrdt.Manager` interface. It will just do nothing and will save us from lots of bugs.

File: cmd/internal/container/mesos/handler.go [hunk 3]
```diff
@@ -23,9 +23,11 @@ import (
 	containerlibcontainer "github.com/google/cadvisor/container/libcontainer"
 	"github.com/google/cadvisor/fs"
 	info "github.com/google/cadvisor/info/v1"
+	"k8s.io/klog"
 
 	cgroupfs "github.com/opencontainers/runc/libcontainer/cgroups/fs"
 	libcontainerconfigs "github.com/opencontainers/runc/libcontainer/configs"
+	resctrl "github.com/opencontainers/runc/libcontainer/intelrdt"
```
Comment (iwankgb) [inline cmd/internal/container/mesos/handler.go]: Why are you aliasing it?

File: info/v1/container.go [hunk 4]
```diff
@@ -842,6 +842,22 @@ type ProcessStats struct {
...
 
+type MemoryBandwidthMonitoringStats struct {
+	// The 'mbm_total_bytes'.
+	TotalBytes uint64 `json:"mbm_total_bytes,omitempty"`
+
+	// The 'mbm_local_bytes'.
+	LocalBytes uint64 `json:"mbm_local_bytes,omitempty"`
+
+	// The 'llc_occupancy'.
+	LLCOccupancy uint64 `json:"llc_occupancy,omitempty"`
```
Comment (iwankgb) [inline info/v1/container.go:887]: It is not memory bandwidth related at all :P

**Review by iwankgb — CHANGES_REQUESTED** [review]
One general and important remark: instead of adding RDT support to each container type you could add it to `manager.manager` and `manager.container`. RDT is just another source of metrics and both these two structs already support NVidia and perf events. See `stats` package for interfaces and no-op implementations :)

File: info/v1/container.go (hunk 4 above)
Comment (Creatone) [reply info/v1/container.go:887]: Yes, but it's part of Memory Bandwidth Monitoring. I decided to add metrics for all MBM features.

File: go.mod (hunk 1 above)
Comment (Creatone) [reply go.mod:35]: I see only the updated one.

Comment (iwankgb) [reply go.mod:35]: And you are absolutely right :)

File: info/v1/container.go (hunk 4 above)
Comment (iwankgb) [reply info/v1/container.go:887]: Last level cache occupancy comes from Cache Monitoring Technology rather than from Memory Bandwidth Monitoring. See: https://01.org/cache-monitoring-technology

File: manager/manager.go [hunk 5]
```diff
@@ -18,6 +18,7 @@ package manager
 import (
 	"flag"
 	"fmt"
+	"github.com/google/cadvisor/resctrl"
```
Comment (iwankgb) [inline manager/manager.go]: Run goimports on the file, please.

File: resctrl/manager.go [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+	stats.NoopDestroy
+}
+
+func (m manager) GetCollector(resctrlPath string) (stats.Collector, error) {
+	collector := newCollector(m.id, resctrlPath)
+	return collector, nil
+}
+
+func NewManager(id string) (stats.Manager, error) {
+	return &manager{id: id}, nil
```
Comment (iwankgb) [inline resctrl/manager.go]: You should rather return instance of `manager` if resctrl filesystem is available or instance of `stats.NoopManager` otherwise.

File: manager/manager.go [hunk 7]
```diff
@@ -961,6 +969,18 @@ func (m *manager) createContainerLocked(containerName string, watchSource watche
 		}
 	}
 
+	if intelrdt.IsMbmEnabled() {
```
Comment (iwankgb) [inline manager/manager.go]: This condition should be handled in `resctrl.NewManager()`.

File: resctrl/collector.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/stats"
+	"github.com/opencontainers/runc/libcontainer/configs"
+	"github.com/opencontainers/runc/libcontainer/intelrdt"
+)
+
+type collector struct {
+	resctrlManager intelrdt.IntelRdtManager
```
Comment (iwankgb) [inline resctrl/collector.go]: Rename this property, please. It will make other developers think that it's an instance of `manager`.

File: resctrl/collector.go [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Collector of resctrl for a container.
+package resctrl
+
+import (
+	info "github.com/google/cadvisor/info/v1"
+	"github.com/google/cadvisor/stats"
+	"github.com/opencontainers/runc/libcontainer/configs"
```
Comment (iwankgb) [inline resctrl/collector.go:24]: There should be empty line between stdlib imports and cadvisor imports, and between cadvisor imports and external imports. Add the line and run goimports, please.

File: resctrl/collector.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+func (c *collector) UpdateStats(stats *info.ContainerStats) error {
+	stats.Resctrl = info.ResctrlStats{}
+
+	resctrlStats, err := c.resctrlManager.GetStats()
+	if err != nil {
+		return err
+	}
+
+	numberOfNUMANodes := len(*resctrlStats.MBMStats)
+	stats.Resctrl.MemoryBandwidthMonitoring = make([]info.MemoryBandwidthMonitoringStats, numberOfNUMANodes)
```
Comment (iwankgb) [inline resctrl/collector.go]:
```suggestion
	stats.Resctrl.MemoryBandwidthMonitoring = make([]info.MemoryBandwidthMonitoringStats, 0, numberOfNUMANodes)
```
such fix will allow you to avoid using index below.

**Review by iwankgb — CHANGES_REQUESTED** [review]

File: metrics/testdata/prometheus_metrics [hunk 11]
```diff
@@ -364,24 +380,122 @@ container_threads_max{container_env_foo_env="prod",container_label_foo_label="ba
 # HELP container_ulimits_soft Soft ulimit values for the container root process. Unlimited if -1, except priority and nice
 # TYPE container_ulimits_soft gauge
 container_ulimits_soft{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",ulimit="max_open_files",zone_name="hello"} 16384 1395066363000
+# HELP go_gc_duration_seconds A summary of the GC invocation durations.
```
Comment (katarzyna-z) [inline metrics/testdata/prometheus_metrics]: Why did you change metrics which are not added in this pull request?

File: metrics/testdata/prometheus_metrics [hunk 12]
```diff
@@ -133,7 +133,23 @@ container_hugetlb_usage_bytes{container_env_foo_env="prod",container_label_foo_l
 container_hugetlb_usage_bytes{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",pagesize="2Mi",zone_name="hello"} 4 1395066363000
 # HELP container_last_seen Last time a container was seen by the exporter
 # TYPE container_last_seen gauge
-container_last_seen{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",zone_name="hello"} 1.395066363e+09 1395066363000
+container_last_seen{container_env_foo_env="prod",container_label_foo_label="bar",id="testcontainer",image="test",name="testcontaineralias",zone_name="hello"} 1.58651404e+09 1586514040481
```
Comment (katarzyna-z) [inline metrics/testdata/prometheus_metrics]: you've edited container_last_seen metric which is not needed in this pr

File: info/v1/container.go (hunk 4 above)
Comment (Creatone) [reply info/v1/container.go:887]: Ok, I see now. I thought that it's part of MBM because of ``/sys/fs/resctrl/info/L3_MON/mon_features``.

File: resctrl/manager.go [hunk 13]
```diff
@@ -19,6 +19,7 @@ package resctrl
 
 import (
 	"github.com/google/cadvisor/stats"
+	"github.com/opencontainers/runc/libcontainer/intelrdt"
```
Comment (iwankgb) [inline resctrl/manager.go:23]: Add space between imports, please.

File: info/v2/conversion.go [hunk 14]
```diff
@@ -154,6 +154,10 @@ func ContainerStatsFromV1(containerName string, spec *v1.ContainerSpec, stats []
 		if len(val.PerfStats) > 0 {
 			stat.PerfStats = val.PerfStats
 		}
+		if len(val.Resctrl.MemoryBandwidthMonitoring) > 0 {
```
Comment (iwankgb) [inline info/v2/conversion.go]: You should update the test case too.

File: info/v1/container.go [hunk 15]
```diff
@@ -874,6 +874,19 @@ type ProcessStats struct {
 	Ulimits []UlimitSpec `json:"ulimits,omitempty"`
 }
 
+type MemoryBandwidthMonitoringStats struct {
+	// The 'mbm_total_bytes'.
+	TotalBytes uint64 `json:"mbm_total_bytes,omitempty"`
+
+	// The 'mbm_local_bytes'.
+	LocalBytes uint64 `json:"mbm_local_bytes,omitempty"`
```
Comment (iwankgb) [inline info/v1/container.go:882]: Have you removed LLC stats completely? Why?

**Review by iwankgb — CHANGES_REQUESTED** [review]

File: info/v1/container.go (hunk 15 above)
Comment (Creatone) [reply info/v1/container.go:882]: I thought that WCA doesn't collect this... I'll add this in a CMT stats.

Comment (iwankgb) [reply info/v1/container.go:882]: We should expose it if we expose RDT based stats.

File: info/v2/conversion_test.go [hunk 16]
```diff
@@ -208,7 +208,20 @@ func TestContainerStatsFromV1(t *testing.T) {
 				Name:         "cycles",
 			},
 		},
+		Resctrl: v1.ResctrlStats{
+			MemoryBandwidthMonitoring: []v1.MemoryBandwidthMonitoringStats{
```
Comment (iwankgb) [inline info/v2/conversion_test.go:212]: What about LLC metrics?

**Review by iwankgb — CHANGES_REQUESTED** [review]

**Comment by iwankgb** [comment]
eldżitiem if all tests pass.

File: manager/container.go [hunk 17]
```diff
@@ -662,6 +668,11 @@ func (c *containerData) updateStats() error {
 		klog.Errorf("error occurred while collecting perf stats for container %s: %s", cInfo.Name, err)
 		return perfStatsErr
 	}
+	if resctrlStatsErr != nil {
+		klog.Errorf("error occurred while collecting perf stats for container %s: %s", cInfo.Name, err)
```
Comment (katarzyna-z) [inline manager/container.go]: incorrect error message, replace 'perf'
